### PR TITLE
Extend symbols with members from implicit conversions/ extensions

### LINKED ISSUE #79: Extend members of class-likes with methods added with extensions / implicit conversions

### PR DESCRIPTION
**Pull request by pikinier20**
closes #79 

Example:
http://scala3doc.s3.eu-central-1.amazonaws.com/pr-194/self/scala3doc/tests/implicitConversions/index.html

### CODE REVIEW

File: src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala [hunk 1]
```diff
@@ -83,6 +84,41 @@ trait ClassLikeSupport:
         }
       }.toList
 
+    def getImplicitConversions: List[ImplicitConversion] =
+      val conversionSymbol = Symbol.requiredClass("scala.Conversion")
+      val givenFields = membersToDocument.collect {
+          case vd: ValDef if vd.symbol.flags.is(Flags.Given) => vd
+        }.toList
+        .filter(_.tpt.tpe.baseClasses.contains(conversionSymbol))
```
Comment (abgruszecki) [inline src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala]: It's better to use `derivesFrom` here.

File: src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala [hunk 2]
```diff
@@ -83,6 +84,41 @@ trait ClassLikeSupport:
...
+        .map(vd => vd.symbol.dri -> vd.tpt.tpe.baseType(conversionSymbol))
+
+      val implicitVals = membersToDocument.collect {
+          case vd: ValDef if vd.symbol.flags.is(Flags.Implicit) => vd
+        }
+        .toList
+        .filter(_.tpt.tpe.baseClasses.contains(conversionSymbol))
```
Comment (abgruszecki) [inline src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala]: Ditto:  It's better to use derivesFrom here.

File: src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala [hunk 3]
```diff
@@ -83,6 +84,41 @@ trait ClassLikeSupport:
...
+        .map(vd => vd.symbol.dri -> vd.tpt.tpe.baseType(conversionSymbol))
+
+      val implicitConversionDefs = getMethods
+        .filter(sym => sym.flags.is(Flags.Implicit) && (sym.paramSymss.size == 0 || (sym.paramSymss.size == 1 && sym.paramSymss(0).size == 0)))
+        .toList
+        .map(_.tree.asInstanceOf[DefDef])
+        .filter(_.returnTpt.tpe.baseClasses.contains(conversionSymbol))
```
Comment (abgruszecki) [inline src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala]: Ditto: It's better to use derivesFrom here.

**Review by abgruszecki — COMMENTED** [review]
Nice piece of work! However:
1) Implicit conversions can only add fields and methods, never types, class(likes). The situation with givens/implicits/extensions is more complicated, but I wouldn't show them either. To keep it short, AFAIK one can reference those by name explicitly, but they don't work as though they were defined inside the class.
2) I'm not certain if we want to propagate givens/implicits/extensions down packages. Consider:
   ```scala
   package Demo:
       class Foo
       class Bar:
           def bar = "bar"
       implicit def foobar(foo: Foo): Bar = Bar()
   ```
   Here, if I import _just_ `Demo.Foo`, I can't call `bar` on it. I'd err on the side of not showing members that can be potentially added, so here `bar` shouldn't be shown in `Foo`. Note that if `Demo` was an _object_, I could call `bar` on an imported `Demo.Foo`, so we should still propagate extensions/givens/implicits down objects.
2) Members added by extensions/givens/implicits defined in companion object of a class are always available and we should show them.
3) Some extensions can be conditional. Example:
   ```scala
    extension (opt: Opt[Int]) def sum: Int = ???
    class Opt[A]
   ```
   Scaladoc _somehow_ manages to write an explanation for when such members are available (see merge method [here](https://www.scala-lang.org/api/current/scala/util/Either.html#merge:A)). This particular explanation seems misleading and the feature in general seems troublesome, so I'm not sure if we want to go that far. Still, I think we should show additional explanation for methods like that. Can one define conditional extension methods in Kotlin? If yes, how are those documented by Dokka?

**Comment by abgruszecki** [comment]
My additional tests (can be placed at the bottom of `implicitConversions`:
```
package nested {
    extension (opt: Opt[Int]) def sum: Int = ???
    class Opt[A]

    class Lst[A]
    object Lst {
        extension (lst: Lst[Int]) def sum: Int = ???
    }

    object Wrapper {
        class Foo
        class Bar {
            def bar = "bar"
        }
        implicit def foobar(foo: Foo): Bar = Bar()
    }

    class Z
}
```

**Comment by abgruszecki** [comment]
Slightly related note: extension method results can reference their "receiver" argument in their result type:
```
scala> extension (s: String) def foo: Option[s.type] = Some(s)
def extension_foo(s: String): Option[s.type]

scala> "a".foo
val res2: Option["a"] = Some(a)
```
It seems like to make things not confusing, we should just display the "receiver" argument of the extension adding a method.

File: src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala (hunk 1 above)
Comment (pikinier20) [reply src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala]: changed

File: src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala (hunk 2 above)
Comment (pikinier20) [reply src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala]: changed

File: src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala (hunk 3 above)
Comment (pikinier20) [reply src/main/scala/dotty/dokka/tasty/ClassLikeSupport.scala]: changed

**Comment by pikinier20** [comment]
I've based it on top of [PR](https://github.com/lampepfl/scala3doc/pull/199) which allow me to add also implicit inherited fields. Implicits from packages are not going down anymore. Also removed implicitly available givens, extensions and types

**Comment by abgruszecki** [comment]
Alright, this looks good now, merging.

**Pull request merged**